validation/link-validator.js

```diff
@@ -0,0 +1,172 @@
+  };
+}
+
+/**
+ * Checks the source and update link lists and compares to find the differences
+ *
+ * @param {list of mdast link nodes} sourceLinks
+ * @param {list of mdast link nodes} updateLinks
+ * @param {path to the file} entry
+ * @returns Log messages for the reporter based on findings
+ */
+export function deepCompare(sourceLinks, updateLinks, path) {
+  console.log(`Deep comparing links on source and updated files at this path: ${path}`);
+  const linkLog = {};
+
+  sourceLinks.forEach((link, index) => {
+    const updateLink = updateLinks[index];
+
+    let sourceUrl;
+    let updateUrl;
+    try {
+      sourceUrl = new URL(link.url);
+    } catch (e) {
+      sourceUrl = false;
+      linkLog[`relative-source-link-${index}`] = link.url;
+      linkLog[`relative-source-text-${index}`] = link?.children[0].value;
+    }
+    try {
+      updateUrl = new URL(updateLink.url);
+    } catch (e) {
+      updateUrl = false;
+      linkLog[`relative-update-link-${index}`] = link.url;
+      linkLog[`relative-update-text-${index}`] = link?.children[0].value;
+    }
+
+    // Partial matches are not fully qualified urls
+    if (!sourceUrl || !updateUrl) {
+      linkLog[`partialUrlMatch-${index}`] = link.url === updateLink.url;
+    }
+
+    linkLog[`sourceLink-${index}`] = link.url;
+    linkLog[`updatedLink-${index}`] = link.children[0].value;
+    linkLog[`sourceText-${index}`] = updateLinks[index].url;
+    linkLog[`updatedText-${index}`] = updateLinks[index]?.children[0]?.value;
+    linkLog[`linksMatch-${index}`] = link.url === updateLink.url;
+    linkLog[`hashMatch-${index}`] = sourceUrl ? sourceUrl.hash === updateUrl.hash : '';
+    linkLog[`hostMatch-${index}`] = sourceUrl ? sourceUrl.host === updateUrl.host : '';
+    linkLog[`pathMatch-${index}`] = sourceUrl ? sourceUrl.pathname === updateUrl.pathname : '';
+    linkLog[`searchMatch-${index}`] = sourceUrl ? sourceUrl.search === updateUrl.search : '';
+    linkLog[`textMatch-${index}`] = link?.children[0]?.value === updateLink?.children[0]?.value;
+  });
+
+  return ['Deep Compare Links', LINKS_DO_NOT_MATCH, path, { log: linkLog }];
+}
+
+/**
+ * Does an initial check for link matching, if it finds issue, it runs the deepCompare function
+ *
+ * @param {list of source links} sourceLinks
+ * @param {list of updated links} updatedLinks
+ * @param {path to files} path
+ * @returns Returns a message for reporter
+ */
+export function compareLinkLists(sourceLinks, updatedLinks, path) {
+  console.log(`Comparing source and update files at this path: ${path}`);
+  // If not the same length, something is wrong from the start
+  if (sourceLinks.length !== updatedLinks.length) {
+    return ['Compare Links', 'list length', LENGTHS_DO_NOT_MATCH];
+  }
+
+  const linksMatch = !sourceLinks.map((link, i) => {
+    const updated = updatedLinks[i];
+    return link.url === updated.url
+    && link?.children[0]?.value === updated?.children[0]?.value;
```

Comment:
If `link` and `updated` might be `undefined`, shouldn't you check in the first half too to avoid a reference error?
```suggestion
    return link?.url === updated?.url
    && link?.children[0]?.value === updated?.children[0]?.value;
```

Reply:
Good point, removing the `?` after `link`